internal/lefthook/run.go

```diff
@@ -10,6 +56,293 @@ func Run(opts *Options, hookName string, gitArgs []string) error {
 }
 
 func (l *Lefthook) Run(hookName string, gitArgs []string) error {
-	// Implement run command
+	cfg, err := config.Load(l.Fs, l.repo.RootPath)
+	if err != nil {
+		return err
+	}
+	if err := cfg.Validate(); err != nil {
+		log.Info(log.Brown(fmt.Sprintf("Config error! %s", err)))
+		return err
+	}
+
+	if tags := os.Getenv("LEFTHOOK_EXCLUDE"); tags != "" {
+		envExcludedTags = strings.Split(tags, ",")
+	}
+
+	var envDisabledOutputs []string
+	if tags := os.Getenv("LEFTHOOK_QUIET"); tags != "" {
+		envDisabledOutputs = strings.Split(tags, ",")
+	}
+	successSprint = outputDisabler(outputSuccess, cfg.SkipOutput, envDisabledOutputs)
+
+	hook, ok := cfg.Hooks[hookName]
+	if !ok {
+		log.Info(log.Brown(fmt.Sprintf("Can't find hook by name '%s'.", hookName)))
+		return errors.New("hook is not found")
+	}
+	if err := hook.Validate(); err != nil {
+		log.Info(log.Brown(fmt.Sprintf("Config error! %s", err)))
+		return err
+	}
+
+	startTime := time.Now()
+
+	if !isOutputDisabled(outputMeta, cfg.SkipOutput, envDisabledOutputs) {
+		log.Info(log.Cyan("Lefthook v" + version.Version()))
+		log.Info(log.Cyan("RUNNING HOOK:"), log.Bold(hookName))
+	}
+
+	spinner.Start()
+
+	var wg sync.WaitGroup
+	l.executeScripts(hookName, gitArgs, cfg.SourceDir, hook, &wg)
+	l.executeScripts(hookName, gitArgs, cfg.SourceDirLocal, hook, &wg)
+
+	for _, name := range hook.CommandsSorted {
+		wg.Add(1)
+
+		if hook.Parallel {
+			go l.executeCommand(name, gitArgs, hook, &wg)
+		} else {
+			l.executeCommand(name, gitArgs, hook, &wg)
+		}
+	}
+
+	wg.Wait()
+	spinner.Stop()
+
+	if !isOutputDisabled(outputSummary, cfg.SkipOutput, envDisabledOutputs) {
+		printSummary(time.Since(startTime))
+	}
+
+	if len(failList) == 0 {
+		return nil
+	}
+	return errors.New("Have failed script")
+}
+
+func (l *Lefthook) executeCommand(commandName string, gitArgs []string, hook *config.Hook, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if hook.Piped && isPipeBroken {
+		spinner.RestartWithMsg("\n", log.Bold(commandName), log.Brown("(SKIP BY BROKEN PIPE)"))
+		return
+	}
+
+	command := hook.Commands[commandName]
+
+	if command.DoSkip(l.repo.State()) {
+		spinner.RestartWithMsg(successSprint(log.Bold(commandName), log.Brown("(SKIP BY SETTINGS)")))
+		return
+	}
+	if isContainsExcludedTags(command.Tags, hook.ExcludeTags) {
+		spinner.RestartWithMsg(successSprint(log.Bold(commandName), log.Brown("(SKIP BY TAGS)")))
+		return
+	}
+
+	runner := command.GetRunner()
+
+	var files []string
+	filesCommand := hook.Files
+	if command.Files != "" {
+		filesCommand = command.Files
+	}
+
+	if strings.Contains(runner, subStagedFiles) {
+		files, _ = l.repo.StagedFiles()
+	} else if strings.Contains(runner, subFiles) || filesCommand != "" {
+		files, _ = git.FilesByCommand(filesCommand)
+	} else if strings.Contains(runner, pushFiles) {
+		files, _ = l.repo.PushFiles()
+	} else {
+		files, _ = l.repo.AllFiles()
+	}
```

Comment:
It looks like we should craft different sets of files for each sub and use them for replacement below.

Reply:
In current implementation `runner` can have only one `{some_files}` template. And it would cause a bug, if we use a few templates here. Although it is hard to imagine a case when we use, say, `{push_files}` and `{staged_files}` in one command, it's better to handle this case properly.